fendermint/vm/topdown/src/sync/syncer.rs

```diff
@@ -0,0 +1,268 @@
+    /// time, since block 3 is a null block, empty data will also be pushed to cache. Block 4 is ready
+    /// to be proposed.
+    pub async fn sync(&mut self) -> anyhow::Result<()> {
+        if self.is_syncing_peer().await? {
+            tracing::debug!("syncing with peer, skip parent finality syncing this round");
+            return Ok(());
+        }
+
+        let chain_head = if let Some(h) = self.finalized_chain_head().await? {
+            h
+        } else {
+            return Ok(());
+        };
+        tracing::debug!(
+            chain_head,
+            pointers = self.sync_pointers.to_string(),
+            "syncing heights"
+        );
+
+        if self.detected_reorg_by_height(chain_head) {
+            tracing::warn!(
+                pointers = self.sync_pointers.to_string(),
+                chain_head,
+                "reorg detected from height"
+            );
+            return self.reset_cache().await;
+        }
+
+        if !self.has_new_blocks(chain_head) {
+            tracing::debug!("the parent has yet to produce a new block");
+            return Ok(());
+        }
+
+        let tail = self.sync_pointers.tail();
+        if let Some((confirmed_height, payload)) = self.poll_next().await? {
+            atomically_or_err::<_, Error, _>(|| {
+                for h in (tail + 1)..confirmed_height {
+                    self.provider.new_parent_view(h, None)?;
+                    tracing::debug!(height = h, "null block pushed to cache");
+                }
+                self.provider
+                    .new_parent_view(confirmed_height, Some(payload.clone()))?;
+                tracing::debug!(height = confirmed_height, "non-null block pushed to cache");
+                Ok(())
+            })
+            .await?;
+        }
+
+        Ok(())
+    }
+}
+
+impl<T, C> LotusParentSyncer<T, C>
+where
+    T: ParentFinalityStateQuery + Send + Sync + 'static,
+    C: tendermint_rpc::Client + Send + Sync + 'static,
+{
+    async fn is_syncing_peer(&self) -> anyhow::Result<bool> {
+        let status: tendermint_rpc::endpoint::status::Response = self
+            .tendermint_client
+            .status()
+            .await
+            .context("failed to get Tendermint status")?;
+        Ok(status.sync_info.catching_up)
+    }
+
+    /// Poll the next block height. Returns finalized and executed block data.
+    async fn poll_next(&mut self) -> Result<Option<(BlockHeight, ParentViewPayload)>, Error> {
+        let height = self.sync_pointers.head() + 1;
+        let parent_block_hash = self.non_null_parent_hash().await;
+
+        let block_hash_res = match self.parent_proxy.get_block_hash(height).await {
+            Ok(res) => res,
+            Err(e) => {
+                let err = e.to_string();
+                if is_null_round_str(&err) {
+                    tracing::warn!(height, "null round at height");
```

Comment:
```suggestion
                    tracing::debug!(height, "null round at height");
```
When emitting a `warn!` think about what the operator of this chain should do to make this warning go away. There is nothing they can do.